src/main/scala/org/allenai/dqa/matching/MatchingModel.scala

```diff
@@ -12,68 +12,159 @@ import edu.cmu.dynet.dynet_swig._
-      matching <- matchRemaining(targetParts.toList, sourceParts.toSet, preprocessing)
+      preprocessing = preprocess(source, target, cg)
+      matching <- matchRemaining(targetParts.toList, sourceParts.toSet, List(), preprocessing)
     } yield {
       MatchingLabel(matching.map(x => (x._1.ind, x._2.ind)).toMap)
     }
   }
 
-  def preprocess(source: Diagram, target: Diagram, cg: ComputationGraph): MatchingPreprocessing = {
+  def preprocess(source: Diagram, target: Diagram, computationGraph: CompGraph): MatchingPreprocessing = {
+    val cg = computationGraph.cg
     val sourceFeatures = source.features.getFeatureMatrix(source.parts, cg)
     val targetFeatures = target.features.getFeatureMatrix(target.parts, cg)
 
-    val weights = parameter(cg, sourceTargetParam)
+    val distanceWeights = parameter(cg, computationGraph.getParameter(DISTANCE_WEIGHTS))
+    val mlpL1W = parameter(cg, computationGraph.getParameter(MLP_L1_W))
+    val mlpL1B = parameter(cg, computationGraph.getParameter(MLP_L1_B))
+    val mlpL2W = parameter(cg, computationGraph.getParameter(MLP_L2_W))
     
     val matchScores = for {
       sourceFeature <- sourceFeatures
     } yield {
       for {
         targetFeature <- targetFeatures
       } yield {
-        // TODO: different scoring function
-        transpose(sourceFeature) * weights * targetFeature
+        // Learn a distance metric
+        val delta = sourceFeature - targetFeature
+        val dist = -1.0 * (transpose(delta) * distanceWeights * delta)
+        
+        // multilayer perceptron
+        val concatenated = concatenate(new ExpressionVector(List(sourceFeature, targetFeature)))
+        val mlpOutput = (mlpL2W * tanh((mlpL1W * concatenated) + mlpL1B)) 
+
+        // mlpOutput
+        dist
+        // -1.0 * dot_product(delta, delta)
+        // input(cg, 0.0f)
       }
     }
 
     new MatchingPreprocessing(sourceFeatures, targetFeatures, matchScores)
   }
 
+  private def getScores(targetPart: Part, remainingArray: Array[Part],
+      currentMatching: List[(Part, Part)], compGraph: CompGraph,
+      preprocessing: MatchingPreprocessing): Expression = {
+    val cg = compGraph.cg
+    val unaryScores = remainingArray.map(x => preprocessing.getMatchScore(x, targetPart))
+    
+    val scores = if (binaryFactors) {
+      val binaryDistanceWeights = parameter(cg,
+          compGraph.getParameter(BINARY_DISTANCE_WEIGHTS))
+      val binaryW = parameter(cg, compGraph.getParameter(BINARY_W))
+      val binaryB = parameter(cg, compGraph.getParameter(BINARY_B))
+      val binaryDistanceWeightsNonlinear = parameter(cg,
+          compGraph.getParameter(BINARY_HIDDEN_DIST))
+          
+      val curTargetFeatures = preprocessing.targetFeatures(targetPart.ind)
+      
+      val binaryScores = for {
+        curSource <- remainingArray
+      } yield {
+        val curSourceFeatures = preprocessing.sourceFeatures(curSource.ind)
+        
+        val pairwiseScores = for {
```

Comment:
similarly, this is a lot of logic to include in a for comprehension, it might be more readable if this were pulled out into a function